rust/agama-cli/src/auth.rs

```diff
@@ -0,0 +1,246 @@
+    /// into an interactive prompt.
+    Login(LoginArgs),
+    /// Release currently stored JWT
+    Logout,
+    /// Prints currently stored JWT to stdout
+    Show,
+}
+
+/// Main entry point called from agama CLI main loop
+pub async fn run(subcommand: AuthCommands) -> anyhow::Result<()> {
+    match subcommand {
+        AuthCommands::Login(options) => login(LoginArgs::proceed(options).password()?).await,
+        AuthCommands::Logout => logout(),
+        AuthCommands::Show => show(),
+    }
+}
+
+/// Reads stored JWT token and returns it
+pub fn jwt() -> anyhow::Result<String> {
+    if let Some(file) = jwt_file() {
+        if let Ok(token) = read_line_from_file(&file.as_path()) {
+            return Ok(token);
+        }
+    }
+
+    return Err(anyhow::anyhow!("JWT not available"));
+}
+
+/// Stores user provided configuration for login command
+#[derive(Args, Debug)]
+pub struct LoginArgs {
+    #[arg(long, short = 'p')]
+    password: Option<String>,
+    #[arg(long, short = 'f')]
+    file: Option<PathBuf>,
+}
+
+impl LoginArgs {
+    /// Transforms user provided options into internal representation
+    /// See Credentials trait
+    fn proceed(options: LoginArgs) -> Box<dyn Credentials> {
+        match options.password {
+            // explicitly provided user + password
+            Some(p) => Box::new(KnownCredentials { password: p }),
+            _ => match options.file {
+                // try to read user + password from a file
+                Some(f) => Box::new(FileCredentials { path: f }),
+                // last instance - ask user to enter user + password interactively
+                _ => Box::new(MissingCredentials {}),
+            },
+        }
+    }
+}
+
+/// Placeholder for no configuration provided by user
+struct MissingCredentials;
+
+/// Stores whatever is needed for reading credentials from a file
+struct FileCredentials {
+    path: PathBuf,
+}
+
+/// Stores credentials as provided by the user directly
+struct KnownCredentials {
+    password: String,
+}
+
+/// Transforms credentials from user's input into format used internaly
+trait Credentials {
+    fn password(&self) -> io::Result<String>;
+}
+
+impl Credentials for KnownCredentials {
+    fn password(&self) -> io::Result<String> {
+        Ok(self.password.clone())
+    }
+}
+
+impl Credentials for FileCredentials {
+    fn password(&self) -> io::Result<String> {
+        read_line_from_file(&self.path.as_path())
+    }
+}
+
+impl Credentials for MissingCredentials {
+    fn password(&self) -> io::Result<String> {
+        let password = read_credential("Password".to_string())?;
+
+        Ok(password)
+    }
+}
+
+/// Path to file where JWT is stored
+fn jwt_file() -> Option<PathBuf> {
+    Some(home::home_dir()?.join(DEFAULT_JWT_FILE))
+}
+
+/// Reads first line from given file
+fn read_line_from_file(path: &Path) -> io::Result<String> {
+    if !&path.exists() {
```

Comment:
No need to use the `&`, it is already a reference.

Reply:
ah yes ... a direct consequence of redoing things multiple times as things evolved ;-)